Preliminary Android build support

mukilan opened the pull request:
The built APK still doesn't run on device and it crashes on loading.
This is happening only after the rust compiler toolchain update. I
am still investigating this, but it is proving to be challenging as
the generated stack trace doesn't show function names even
in debug build and after ensuring Gradle doesn't strip the symbols.

The goal with this PR is to land the changes with just the CI build
passing.

Other issues that need to be resolved in future PRs
1. Enable GStreamer support - this is disabled for now
2. Enable bluetooth support - blurdroid is unmaintained and doesn't
    compile now.
3. Revisit issues involving dependencies - jemalloc-sys, freetype and
    fontconfig
4. Revisit compositing logic & event handling. Apps can lose underlying
    GL Surface when they are sent to background. I have added logic to
    pause/resume the compositor when this happens so that we bind/unbind.
    Scrolling is also now 'janky' after the move from WebRender based zoom
    to Servo handling the zoom with root pipeline.



---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [x] These changes do not require tests because these changes are for preliminary support. CI sanity check of android is disabled for now as apk crashes and sanity test code is also incorrect.

mrobinson committed:
Android build

* Fixes
* More fixes
  - Still failing in the linking step
* More work on getting linking working

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
android: use mozjs with ndk r25c. loads servo.org

more android build fixes.

mukilan committed:
fix ./mach run for android and make it follow logs

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
add experimental logic for compositor pause/resume

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
pass DPI from android to simpleservo

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
ci: add android workflow

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
switch to ANDROID_SDK_ROOT and ANDROID_NDK_ROOT vars

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
upgrade gradle to 4.10.1

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
upgrade to gradle 5.1.1

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
upgrade to gradle 8 and agp 8

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
make compositing work again with external present

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
android: improve mach support for non-NixOS and CI

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
fix sampler compilation bug introduced in #30490

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
ci: add android build to main workflow

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
gradle: set MinSdk = targetSdk = 30

NDK requires we compile against the minSdk API level
which is 30 in our case.

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
add instructions for android in README.md

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
apk: move servosurface to servoview

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
apk: uncomment the mediasession callbacks on MainActivity

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
apk: fix crash on MainAtivity.onDestroy

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
apk: drop VR, arm 5 and unused code

This commit drops:
* support for google, oculusvr
* support for arm5 architecture

and also removes
* fakeld scripts
* unused java code

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
cleanup shell.nix

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
android: add FIXMEs for gstreamer code

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
apk: remove commented code and debug logs

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
cleanup ServoView.java

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
mach: comment call to download gstreamer deps for android

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
disable bluetooth for jniapi as blurdroid is broken

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
fixup! README.md

mukilan committed:
fixup! remove change in Cargo.toml

mukilan committed:
fixup! move shell variables together

mrobinson commented on ports/libsimpleservo/jniapi/build.rs:
```diff
@@ -3,31 +3,20 @@
         env::var("OUT_DIR").expect("Cargo should have set the OUT_DIR environment variable");
 
-    println!("cargo:rustc-link-lib=static=android_native_app_glue");
+    // FIXME: we need this workaround since jemalloc-sys still links
+    // to libgcc instead of libunwind, but Android NDK 23c and above
+    // don't have libgcc. Alternatively, we could disable jemalloc
+    // for android in servo_allocator
+    let mut libgcc = File::create(out_dir.clone() + "/libgcc.a").unwrap();
+    libgcc.write_all(b"INPUT(-lunwind)").unwrap();
+
```
I think it would simplify the build to just disable jemalloc on Android.

mukilan replied:
The comment is outdated. I did try disabling jemalloc but that requires providing an implementation of [this](https://github.com/servo/servo/blob/main/components/allocator/lib.rs#L39) method. I couldn't find a way to do that with the standard allocator.

mrobinson replied:
Could you use `malloc_usable_size`? https://docs.rs/libc/latest/libc/fn.malloc_usable_size.html

mukilan replied:
I did look into that as well, but the documentation [warns](https://doc.rust-lang.org/std/alloc/struct.System.html) against directly using the backend allocator methods, but I can switch to `malloc_usable_size` we think it should be safe in this specific case.

mrobinson replied:
Yeah, let's just use the backend allocator methods and disable jemaloc. I think it's fine in this case.

mrobinson commented on ports/libsimpleservo/jniapi/src/lib.rs:
```diff
@@ -5,28 +5,44 @@
-use android_logger::{self, Filter};
-use gstreamer::debug_set_threshold_from_string;
+use android_logger::{self, FilterBuilder, Config};
+// FIXME: GStreamer support
+// use gstreamer::debug_set_threshold_from_string;
```
Instead of commenting out the code, please remove it.

mrobinson commented on ports/libsimpleservo/jniapi/src/lib.rs:
```diff
@@ -79,26 +96,31 @@ pub fn Java_org_mozilla_servoview_JNIServo_init(
-        if let Some(gst_debug_str) = gst_debug_str {
-            debug_set_threshold_from_string(&gst_debug_str, true);
-        }
+        // FIXME: GStreamer support
+        // if let Some(gst_debug_str) = gst_debug_str {
+        //   debug_set_threshold_from_string(&gst_debug_str, true);
+        // }
```
Please just remove this code instead of commenting it out.

mrobinson commented on python/servo/build_commands.py:
```diff
@@ -123,7 +123,9 @@ def build(self, build_type: BuildType, jobs=None, params=None, no_package=False,
                 print((key, env[key]))
 
-        self.download_and_build_android_dependencies_if_needed(env)
+        # FIXME: disabled until GStreamer support is enabled
+        # self.download_and_build_android_dependencies_if_needed(env)
+
```
Let's just remove this for now as well as `download_and_build_android_dependencies_if_needed()`.

mrobinson commented on python/servo/command_base.py:
```diff
@@ -618,48 +605,31 @@ def build_android_env_if_needed(self, env: Dict[str, str]):
-        gcc_toolchain = path.join(env['ANDROID_NDK'], "toolchains",
-                                  android_toolchain_prefix + "-4.9", "prebuilt", host)
-        gcc_libs = path.join(gcc_toolchain, "lib", "gcc", android_toolchain_name, "4.9.x")
+        host_cc = env.get('HOST_CC') or shutil.which("clang") or shutil.which(["gcc"])
+        host_cxx = env.get('HOST_CXX') or shutil.which("clang++") or shutil.which("g++")
```
I think we should remove the `gcc` and `g++` fallbacks here, as I don't know if Servo supports gcc any longer.

mrobinson commented on python/servo/command_base.py:
```diff
@@ -671,52 +641,28 @@ def build_android_env_if_needed(self, env: Dict[str, str]):
+
+        # needed by mozjs
+        env['ANDROID_API_LEVEL'] = android_api
+        env["ANDROID_NDK_HOME"] = env["ANDROID_NDK_ROOT"]
```
Hrm. mozjs should probably switch to using the same variable as Servo or vice versa? Which is more common? It looks like both are used?

mukilan replied:
`ANDROID_NDK_ROOT` is  the official one now, so mozjs should be changed. Can that land separately?

mrobinson replied:
Sure. I can make the fix in mozjs.

mrobinson commented on python/servo/command_base.py:
```diff
@@ -671,52 +641,28 @@ def build_android_env_if_needed(self, env: Dict[str, str]):
+        env['CFLAGS'] = "--target=" + android_toolchain_name
+        env['CXXFLAGS'] = "--target=" + android_toolchain_name
+
+        # needed by mozjs
```
```suggestion
        # These two variables are needed for the mozjs compilation.
```

mrobinson commented on python/servo/command_base.py:
```diff
@@ -671,52 +641,28 @@ def build_android_env_if_needed(self, env: Dict[str, str]):
+        env["ANDROID_NDK_HOME"] = env["ANDROID_NDK_ROOT"]
+
+        # The two vars set below are passed by our custom
+        # support/android/toolchain.cmake to the NDK's cmake toolchain file
```
```suggestion
        # support/android/toolchain.cmake to the NDK's CMake toolchain file.
```

mrobinson commented on python/servo/command_base.py:
```diff
@@ -964,21 +906,18 @@ def run_cargo_build_like_command(
-        return path.join(self.context.topdir, "support", "android")
-
-    def android_aar_dir(self):
-        return path.join(self.context.topdir, "target", "android", "aar")
+        return
```
If this is unused, please just remove it.

mrobinson commented on servobuild.example:
```diff
@@ -56,8 +56,8 @@ media-stack = "auto"
 #ndk = "/opt/android-ndk"
-#toolchain = "/opt/android-toolchain"
-#platform = "android-18"
+#toolchain = "/opt/android-toolchain" FIXME: still needed?
+#platform = "android-18" FIXME: still needed?
```
If these aren't read, let's just remove them.

mrobinson commented on support/android/apk/gradlew:
```diff
@@ -150,11 +154,19 @@ if $cygwin ; then
-JVM_OPTS[${#JVM_OPTS[*]}]="-Dorg.gradle.appname=$APP_BASE_NAME"
+APP_ARGS=$(save "$@")
+
+# Collect all arguments for the java command, following the shell quoting and substitution rules
+eval set -- $DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS "\"-Dorg.gradle.appname=$APP_BASE_NAME\"" -classpath "\"$CLASSPATH\"" org.gradle.wrapper.GradleWrapperMain "$APP_ARGS"
+
+# by default we should be in the correct project dir, but when run from Finder on Mac, the cwd is wrong
+if [ "$(uname)" = "Darwin" ] && [ "$HOME" = "$PWD" ]; then
+  cd "$(dirname "$0")"
+fi
```
Is this code all from Gradle?

mukilan replied:
Yep, this generated by Gradle to bootstrap itself on any system.

mrobinson commented on support/android/apk/jni/Android.mk:
```diff
@@ -16,13 +16,13 @@
-LOCAL_MODULE := gstreamer
-LOCAL_SRC_FILES := libgstreamer_android.so
-include $(PREBUILT_SHARED_LIBRARY)
+# include $(CLEAR_VARS)
+# LOCAL_PATH:= $(SERVO_TARGET_DIR)/../../gstreamer/gst-build-$(APP_ABI)
+# LOCAL_MODULE := gstreamer
+# LOCAL_SRC_FILES := libgstreamer_android.so
+# include $(PREBUILT_SHARED_LIBRARY)
```
Let's just remove this for now.

mrobinson commented on support/android/apk/jni/Application.mk:
```diff
@@ -1,5 +1,6 @@
-APP_STL:= c++_shared
-APP_ABI:= armeabi-v7a x86
+APP_MODULES := c++_shared servojni
+#APP_MODULES := c++_shared servojni #gstreamer
```
Ditto.

mrobinson commented on support/android/apk/servoapp/src/main/java/org/mozilla/servo/MainActivity.java:
```diff
@@ -78,8 +78,8 @@ protected void onCreate(Bundle savedInstanceState) {
         String log = intent.getStringExtra("servolog");
-        String gstdebug = intent.getStringExtra("gstdebug");
-        mServoView.setServoArgs(args, log, gstdebug);
+        // TODO String gstdebug = intent.getStringExtra("gstdebug");
```
Ditto.

mrobinson commented on support/android/apk/servoview/build.gradle:
```diff
@@ -145,34 +108,51 @@ android {
+    // integration with native C/C++ shared objects (based on the
+    // `android.externalNativeBuild` dsl object) assumes that we
+    // actually execute compiler commands to produce the shared
+    // objects. We already have a libsimpleservo.so produce by rustc.
```
```suggestion
    // objects. We already have a libsimpleservo.so produced by rustc.
```

mrobinson commented on support/android/apk/servoview/build.gradle:
```diff
@@ -145,34 +108,51 @@ android {
     tasks.all {
         compileTask ->
-            Pattern pattern = Pattern.compile(/^compile[A-Z][\w\d]+([A-Z][\w\d]+)(Debug|Release)Ndk/)
+            // match task `mergeBasicArmv7DebugJniLibFolders`
```
```suggestion
            // This matches task `mergeBasicArmv7DebugJniLibFolders`.
```

mrobinson commented on support/android/apk/servoview/build.gradle:
```diff
@@ -145,34 +108,51 @@ android {
+                    'APP_BUILD_SCRIPT=../jni/Android.mk',
+                    'NDK_APPLICATION_MK=../jni/Application.mk',
+                    'NDK_LIBS_OUT=' + getJniLibsPath(debug, arch),
+                    //'NDK_OUT=' + getJniLibsPath(debug, arch),
```
Please remove.

mrobinson commented on support/android/apk/servoview/src/main/java/org/mozilla/servoview/JNIServo.java:
```diff
@@ -6,21 +6,22 @@
     JNIServo() {
         System.loadLibrary("c++_shared");
-        System.loadLibrary("gstreamer_android");
+        // FIXME: GStreamer support
+        // System.loadLibrary("gstreamer_android");
```
Please just remove this.

mrobinson commented on support/android/apk/servoview/src/main/java/org/mozilla/servoview/Servo.java:
```diff
@@ -30,18 +31,20 @@ public Servo(
 
         try {
-          GStreamer.init((Context) activity);
+          // FIXME: GStreamer support
+          // GStreamer.init((Context) activity);
```
Ditto.

mrobinson commented on support/android/apk/servoview/src/main/java/org/mozilla/servoview/ServoView.java:
```diff
@@ -154,51 +126,9 @@ public void animationStateChanged(boolean animating) {
 
     private void startLooping() {
-      // In case we were already drawing.
+      //In case we were already drawing.
```
I think this change is unintentional.

mrobinson commented on third_party/webrender/webrender/Cargo.toml:
```diff
@@ -9,7 +9,7 @@ build = "build.rs"
 
 [features]
-default = ["freetype-lib"]
+default = ["freetype-lib", "no_static_freetype"]
```
This feature should probably not be added unconditionally, but added via a platform-specific dependency in the compositor's Cargo.toml.

mrobinson commented on .github/workflows/android.yml:
```diff
@@ -0,0 +1,91 @@
+        run: |
+          export JAVA_HOME="${JAVA_HOME_17_X64}"
+          python3 ./mach build --android --target ${{ matrix.arch }} --${{ inputs.profile }}
+      # TODO: fix test-android-startup
+      # - name: Script tests
+      #   run: ./mach test-android-startup
```
Would it be possible to open a bug for this issue and link here?

```suggestion
      # TODO: Android currently crashes on startup, so this is disabled currently.
      # https://github.com/servo/servo/issue...
      # - name: Script tests
      #   run: ./mach test-android-startup
```

mrobinson commented on .github/workflows/android.yml:
```diff
@@ -0,0 +1,91 @@
+        with:
+          crate: taplo-cli
+          locked: true
+      - name: Bootstrap Python
+        run: python3 -m pip install --upgrade pip virtualenv
+      - name: Bootstrap dependencies
+        run: sudo apt update && python3 ./mach bootstrap
+      - name: Build (arch ${{ matrix.arch }} profile ${{ inputs.profile }})
+        run: |
+          export JAVA_HOME="${JAVA_HOME_17_X64}"
```
Don't we need to install a specific version of the SDK and NDK here?

mukilan replied:
We are relying on the SDK and NDK preinstalled in the [runner image](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md#environment-variables-2)

mrobinson replied:
Let's use actions like "nttld/setup-ndk" that we use for mozjs to ensure that the proper version is installed. If the runner image changes, this image this might break one day in the future.

mrobinson commented on README.md:
```diff
@@ -116,6 +116,27 @@ though of course it doesn’t produce a binary you can run.
 ### Building for Android target
 
+Prerequisites:
+Servo's build system assumes that both the Android SDK & NDK are
```
It'd be great to have a small set of Android-specific instructions here like we have for other platforms. I'm happy to write this.

mukilan replied:
I am happy to write this.  Also, there is a link at the top of this README to servo wiki page with more [detailed instructions](https://github.com/servo/servo/wiki/Building-for-Android), but this is out of date. Do we want to add Android-specific instructions here or keep them in the wiki?

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -501,6 +501,26 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+    #[allow(unsafe_code)]
+    pub fn resume(&mut self, native_widget: *mut c_void, coords: DeviceIntSize) {
+        debug!("Resuming compositor: {native_widget:?}");
+        // Reinitialize webrender surfman and bind to context
```
Can you expand this comment into a full rustdoc for this function? I'm not entirely sure why this new functionality is required in the compositor.

mukilan replied:
Added rustdoc comment on the renamed method `replaced_native_surface`

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -501,6 +501,26 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
         self.shutdown_state = ShutdownState::FinishedShuttingDown;
     }
 
+    pub fn pause(&mut self) {
```
It might be nice to have some rustdoc here on this function.

mrobinson commented on components/compositing/windowing.rs:
```diff
@@ -105,6 +106,10 @@ pub enum EmbedderEvent {
     ChangeBrowserVisibility(TopLevelBrowsingContextId, bool),
     /// Virtual keyboard was dismissed
     IMEDismissed,
+    /// Pause the compositor when underlying native surface is lost
```
```suggestion
    /// Pause the compositor when underlying native surface is lost.
```

mrobinson commented on components/compositing/windowing.rs:
```diff
@@ -105,6 +106,10 @@ pub enum EmbedderEvent {
     IMEDismissed,
+    /// Pause the compositor when underlying native surface is lost
+    PauseCompositor,
+    /// Resume the compositor when underlying native surface is regained
```
```suggestion
    /// Resume the compositor when underlying native surface is regained.
```

mrobinson commented on components/compositing/windowing.rs:
```diff
@@ -105,6 +106,10 @@ pub enum EmbedderEvent {
     /// Virtual keyboard was dismissed
     IMEDismissed,
+    /// Pause the compositor when underlying native surface is lost
+    PauseCompositor,
```
It might be worth renaming these to make it clearer what these are handling. Perhaps something like:

```
InvalidateNativeSurface,
```

mukilan replied:
FWIW, I wasn't sure about the correctness of the current way of handling this and suspected that eventually this will include more work such as pausing the script thread, animation etc in addition to binding/unbinding surface. But I'll change it to the suggested names for now and we can rename it later.

mrobinson commented on components/compositing/windowing.rs:
```diff
@@ -105,6 +106,10 @@ pub enum EmbedderEvent {
+    /// Pause the compositor when underlying native surface is lost
+    PauseCompositor,
+    /// Resume the compositor when underlying native surface is regained
+    ResumeCompositor(*mut c_void, DeviceIntSize)
```
This could be something like:

```
ReplaceNativeSurface(*mut c_void, DeviceIntSize)
```

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -501,6 +501,26 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+    pub fn resume(&mut self, native_widget: *mut c_void, coords: DeviceIntSize) {
+        debug!("Resuming compositor: {native_widget:?}");
+        // Reinitialize webrender surfman and bind to context
+        let connection = self.webrender_surfman.connection();
+        let native_widget = unsafe {
+            connection.create_native_widget_from_ptr(native_widget, coords.to_untyped())
+        };
+        if let Err(e) = self.webrender_surfman.bind_native_surface_to_context(native_widget) {
+            warn!("Binding native surface to context failed ({:?})", e);
+        }
+    }
+
```
Hrm...so normally the embedder is responsible for binding the native widget to the context, right? Is there a way to rework this so that this always happens in the embedder instead of making this happen in WebRender. Perhaps we could add a method to unset `webrender_surfman` in the compositor and then reset it from the embedder?

mukilan replied:
I'm not sure if I fully understand the suggestion. 

> instead of making this happen in WebRender.

Do you mean in WebRenderSurfman ?

Today, AFAICT, the embedder doesn't  bind the widget directly, but it simply does `WebrenderSurfman::create` during initialization and then holds on to that instance for the lifetime. Are you suggesting that we move away from that model and make embedder recreate `WebrenderSurfman` on demand? Today  `WebrenderSurfman::create` also creates other resources like handle to the device, opengl context and swapchain - are we fine with recreating all these other resources as well?

Also, today the compositor  holds on the same WebrenderSurfman instance that the embedder creates (The data is wrapped in Rc and WindowMethods trait allows libservo to create a new handle from an open window). If we reset webrender_surfman in embedder, wouldn't we also need to reset it from compositor?

mrobinson replied:
I see. Thanks for the explanation! I think this approach is okay for now! We probably need to take a holistic look at how all these surfaces are managed at some point in the future.

mrobinson commented on ports/libsimpleservo/jniapi/Cargo.toml:
```diff
@@ -14,21 +14,27 @@ test = false
-android_logger = "0.10"
-gstreamer = { workspace = true }
+android_logger = "0.13"
+# gstreamer = { workspace = true }
```
Please remove this.

mrobinson commented on ports/libsimpleservo/jniapi/Cargo.toml:
```diff
@@ -14,21 +14,27 @@ test = false
 serde_json = { workspace = true }
-simpleservo = { path = "../api" }
+
+# TODO: enable native-bluetooth feature for android
+# and remove explicit feature list here
+[dependencies.simpleservo]
+path = "../api"
+default-features = false
+features = ["max_log_level", "webdriver"]
```
I think it makes sense to keep using the one line dependency format, which we use pretty consistently in Servo.

mrobinson commented on ports/libsimpleservo/jniapi/Cargo.toml:
```diff
@@ -14,21 +14,27 @@ test = false
 serde_json = { workspace = true }
-simpleservo = { path = "../api" }
+
+# TODO: enable native-bluetooth feature for android
+# and remove explicit feature list here
```
```suggestion
# TODO: Once the native-bluetooth feature works for Android, remove the explicit feature list here.
```

mrobinson commented on ports/libsimpleservo/jniapi/Cargo.toml:
```diff
@@ -14,21 +14,27 @@ test = false
 [features]
 debugmozjs = ["simpleservo/debugmozjs"]
-default = ["max_log_level", "native-bluetooth", "webdriver"]
+# TODO: enable native-bluetooth feature for android
```
I think it's fine to remove this comment.

mrobinson reviewed (changes requested):
Great work! I'm so glad to see this happening. I have a bunch of comments below, but this is already looking very nice.

mukilan committed:
fixup! cleanup jniapi/Cargo.toml comments

mukilan committed:
delete commented gstreamer related android code

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
remove unused config variable in servbuild

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
android: more cleanup

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
force no_static_freetype only for android

mukilan committed:
use actions to manage sdk, ndk and java

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
rename embedder event names to be more clear.

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
link to startup crash issue

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
fix lint issues

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
upgrade env_logger to 0.10 with duplicate exception

libservo and android_logger can use env_logger 0.10
but quickcheck is still stuck on 0.8 and has not seen
any activity in the last 2 years. This commit adds
a duplicate exception until the quickcheck dependency
can be upgraded (or replaced)

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mrobinson commented on .github/workflows/android.yml:
```diff
@@ -0,0 +1,107 @@
+    name: Result
+    runs-on: ubuntu-latest
+    if: always()
+    # needs all build to detect cancellation
```
```suggestion
    # `needs: 'build'` is necessary to detect cancellation.
```

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -501,6 +501,33 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
     }
 
+    /// The underlying native surface can be lost during servo's lifetime.
+    /// On android, for example, this happens when the app is sent to background.
```
```suggestion
    /// On Android, for example, this happens when the app is sent to background.
```

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -501,6 +501,33 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+        }
+    }
+
+    /// On android, this function will be called when the app moves to foreground
```
```suggestion
    /// On Android, this function will be called when the app moves to foreground
```

mrobinson commented on components/compositing/windowing.rs:
```diff
@@ -105,6 +106,14 @@ pub enum EmbedderEvent {
     ChangeBrowserVisibility(TopLevelBrowsingContextId, bool),
     /// Virtual keyboard was dismissed
     IMEDismissed,
+    /// Sent on platforms like android where the native widget surface can be
```
```suggestion
    /// Sent on platforms like Android where the native widget surface can be
```

mrobinson commented on components/compositing/windowing.rs:
```diff
@@ -105,6 +106,14 @@ pub enum EmbedderEvent {
+    /// automatically destroyed by the system, for example when the app
+    /// is sent to background.
+    InvalidateNativeSurface,
+    /// Sent on platforms like android where system recreates a new surface for
```
```suggestion
    /// Sent on platforms like Android where system recreates a new surface for
```

mrobinson commented on components/servo/lib.rs:
```diff
@@ -1091,6 +1098,9 @@ fn default_user_agent_string_for(agent: UserAgent) -> &'static str {
     const DESKTOP_UA_STRING: &'static str =
         "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Servo/1.0 Firefox/111.0";
 
+    #[cfg(target_os = "android")]
```
It might be a good moment to clean this up and just conditionally define a `UA_STRING` and always return it.

mukilan replied:
Hmm, the existing code is written so that the user can switch user agent to ios/android specific ones even on desktop at runtime using [this](https://github.com/servo/servo/blob/main/ports/servoshell/main2.rs#L38) flag, probably to allow testing.

I think implementing what you suggested would require removing this functionality from libservo (maybe we can do it to ports/servoshell). Can this be tackled in a separate PR?

mrobinson replied:
Yes, sure!

mrobinson commented on python/servo/command_base.py:
```diff
@@ -671,52 +636,28 @@ def build_android_env_if_needed(self, env: Dict[str, str]):
+        env["ANDROID_NDK_HOME"] = env["ANDROID_NDK_ROOT"]
+
+        # The two variables set below are passed by our custom
+        # support/android/toolchain.cmake to the NDK's Cmake toolchain file
```
```suggestion
        # support/android/toolchain.cmake to the NDK's CMake toolchain file
```

mrobinson reviewed (approved):
Nice! So happy to see this finally landing. Just a couple small things before landing and also let's go ahead and disable jemalloc for Android.

mukilan committed:
android: fix comments

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
disable jemalloc on android

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
fixup! replace linux with android in cfg

mukilan commented:
Thanks for the review!
